lib/mongodb.js

```diff
@@ -1881,14 +1881,30 @@ MongoDB.prototype.isObjectIDProperty = function(model, prop, value, options) {
   ) {
     return true;
   } else if ('string' === typeof value) {
+    // dataType is specified as ObjectID
+    if (prop && prop.mongodb && prop.mongodb.dataType === ObjectID) {
+      // value looks like an ObjectID, and it should be treated as one
+      // because `prop.mongodb.dataType` is set to 'ObjectID'
+      // this condition overrides the `strictObjectIDCoercion` setting
+      if (value.match(/^[0-9a-fA-F]{24}$/)) return true;
+      // if dataType is specified as ObjectID, value is expected to an ObjectID-like string
+      else throw new Error(`${value} is not an ObjectID string`);
```

Comment:
IMO, this error should be thrown by the code performing coercion from string to ObjectID instance, not by the helper checking whether a property expects ObjectID values.

`isObjectIDProperty` should simply return "true" when the property value should be stored as ObjectID and "false" when not, regardless of whether the actual value can be converted to ObjectID or not.

For example, clients can provide a numeric value for an ObjectID property. Such request will not go through this line, and yet we want to reject it as invalid. BTW, please make sure there is a test case for this scenario.